cpp/src/io/csv/csv_gpu.cu

```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+                                                                           size_t data_size,
+                                                                           size_t byte_range_start,
+                                                                           size_t byte_range_end,
+                                                                           size_t skip_rows,
+                                                                           size_t num_row_offsets,
+                                                                           int terminator,
+                                                                           int delimiter,
+                                                                           int delimiter_esc,
+                                                                           int quotechar,
+                                                                           int escapechar,
+                                                                           int commentchar,
+                                                                           bool skipblanklines)
+{
+  __shared__ __align__(8) uint64_t ctxtree[rowofs_block_dim * 2];
+
+  const char *end = start + (min(parse_pos + chunk_size, data_size) - start_offset);
+  uint32_t t      = threadIdx.x;
+  size_t block_pos =
+    (parse_pos - start_offset) + blockIdx.x * static_cast<size_t>(rowofs_block_bytes) + t * 32;
+  const char *cur = start + block_pos;
+  uint64_t ctxb;
+  uint4 ctx_map;
+  int c, c_prev;
+
+  // Initial state is neutral context, zero rows
+  ctx_map.x = ctx_map.y = ctx_map.z = 0;
+  ctx_map.w                         = (0 << 0) | (1 << 2) | (2 << 4) | (3 << 6);  // i << (2*i)
+  c_prev                            = (cur > start) ? cur[-1] : terminator;
+  // Loop through all 32 bytes and keep a bitmask of row starts for each possible input context
+  for (uint32_t pos = 0; pos < 32; pos++, cur++, c_prev = c) {
+    uint32_t ctx;
+    if (cur < end) {
+      c = cur[0];
+      if (c_prev == terminator) {
+        if (c == commentchar ||
+            (skipblanklines && (c == terminator || (c == '\r' && terminator == '\n')))) {
+          // Start of a new comment or empty row
+          ctx = make_short_row_context(ROW_CTX_COMMENT, ROW_CTX_QUOTE, ROW_CTX_COMMENT, 0, 0, 0);
+        } else if (c == quotechar) {
+          // Quoted string on newrow, or quoted string ending in terminator
+          ctx = make_short_row_context(ROW_CTX_QUOTE, ROW_CTX_NONE, ROW_CTX_QUOTE, 1, 0, 1);
+        } else {
+          // Start of a new row unless within a quote
+          ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_QUOTE, ROW_CTX_NONE, 1, 0, 1);
+        }
+      } else if (c == quotechar) {
+        if (c_prev == delimiter || c_prev == quotechar) {
+          // Quoted string after delimiter, quoted string ending in delimiter, or double-quote
```

Comment:
Have the pandas style double quotes already been replaced by single quotes at this point?

Reply:
No. Basically when there is a double-quote inside a string, it's treated as two back-to-back strings here (leaves then re-enters quote state). If strings are not quoted, then there is no quotechar. Only allowing opening quotes when following a delimiter somewhat improves pandas compatibility for situations when quoting is incorrectly enabled (happens often since quoting is enabled by default -> in this case, we still break when there is a doublequote inside a non-quoted field, but it's presumably more rare)

Reply:
Oh, the double-quote issue is when there is a pair of quotes in the middle of a string, like `"foo""bar"` getting parsed as a single string? I thought it was that strings could have double quotes around them, like `""foobar""`.
Ok, and then `foo""bar` would be parsed incorrectly, but this seems to be invalid input anyway?

Reply:
My understanding is that `"foo""bar"` gets parsed as the single-string string `foo"bar` when double-quotes are enabled. The data string `"foobar"` would be encoded in csv as `"""foobar"""` if quoting and doublequotes are enabled. I'm not entirely sure what happens if doublequotes are not enabled and string fields have quotes in them -> maybe the parser just looks for the next terminator ? (I suspect we'll need to have a lookahead loop to deal with such corner cases). `foo""bar`, `foo "bar" foo` and `11"W x 13"H x 10"D` would all be invalid if quoting is enabled, but I think the last two are more common (I think pandas automatically switches off quoting if the field doesn't start with a quote, so we may just need something like an additional quoted-field info bits if we want to avoid lookahead, at the cost of a slightly more complex merge op).

Reply:
[Let's save these discussions for upcoming pandas compatibility work]